frontend/dockerfile/dockerfile_test.go

```diff
@@ -3390,6 +3392,87 @@ COPY --from=build /dest /dest
 	require.Equal(t, []byte("0644\n0755\n0413\n"), dt)
 }
 
+func testAddURLChecksum(t *testing.T, sb integration.Sandbox) {
+	f := getFrontend(t, sb)
+	f.RequiresBuildctl(t)
+
+	resp := httpserver.Response{
+		Etag:    identity.NewID(),
+		Content: []byte("content1"),
+	}
+	server := httpserver.NewTestServer(map[string]httpserver.Response{
+		"/foo": resp,
+	})
+	defer server.Close()
+
+	c, err := client.New(sb.Context(), sb.Address())
+	require.NoError(t, err)
+	defer c.Close()
+
+	t.Run("Valid", func(t *testing.T) {
+		dockerfile := []byte(fmt.Sprintf(`
+FROM scratch
+ADD --checksum=%[2]s %[1]s /tmp/foo
+`, server.URL+"/foo", digest.FromBytes(resp.Content).String()))
+
+		dir, err := integration.Tmpdir(
+			t,
+			fstest.CreateFile("Dockerfile", dockerfile, 0600),
+		)
+		require.NoError(t, err)
+
+		_, err = f.Solve(sb.Context(), c, client.SolveOpt{
+			LocalDirs: map[string]string{
+				builder.DefaultLocalNameDockerfile: dir,
+				builder.DefaultLocalNameContext:    dir,
+			},
+		}, nil)
+		require.NoError(t, err)
+	})
+	t.Run("DigestMismatch", func(t *testing.T) {
```

Comment:
I recall go-digest panics in some code-paths where an algorithm isn't supported / not registered. Not sure if we can hit that code-path, but perhaps we should add a test;

- with a known, but (currently) unsupported algorithm ( `md5:deadcafe` e.g.)
- with a bogus algorithm (`foo:deadcafe`)
- without a `algorithm:` prefix (`deadcafe`)

Reply:
I recall it has been fixed some times ago https://github.com/opencontainers/go-digest/commit/99f8dd8415c586d4b74b441f5d4ec0a22d00e8e0 so parsing will fist check algorithm availability and then validates: https://github.com/opencontainers/go-digest/blob/bde1400a84bededaf764144a72cbd007a633196e/digest.go#L110-L116

Reply:
Yeah, either way I don't think it would hurt to have a test for invalid or unsupported hashes (also to see if they produce a reasonably useful error message).

Reply:
Added the tests

Reply:
Thank you!